services/supervisord/supervisord.go

```diff
@@ -406,6 +422,39 @@ func (s *Service) marshalConfig(tmpl *template.Template, settings *models.Settin
 	return b, nil
 }
 
+// addAlertManagerParams parses alertManagerURL
+// and extracts url, username and password from it to templateParams.
+func addAlertManagerParams(alertManagerURL string, templateParams map[string]interface{}) error {
+	templateParams["AlertmanagerURL"] = alertManagerURL
+	templateParams["AlertManagerUser"] = ""
+	templateParams["AlertManagerPassword"] = ""
+	if alertManagerURL == "" {
+		return nil
+	}
+	u, err := url.Parse(alertManagerURL)
+	if err != nil {
+		return errors.Wrap(err, "cannot parse AlertManagerURL")
+	}
+	if u.Opaque != "" || u.Host == "" {
+		return errors.Errorf("AlertmanagerURL parsed incorrectly as %#v", u)
+	}
+	if username := u.User.Username(); username != "" {
+		password, _ := u.User.Password()
+		n := url.URL{
+			Scheme:   u.Scheme,
+			Host:     u.Host,
+			Path:     u.Path,
+			RawQuery: u.RawQuery,
+			Fragment: u.Fragment,
+		}
+		templateParams["AlertManagerUser"] = username
```

Comment:
looks like these two variables(`AlertManagerUser`, `AlertManagerPassword`)  can be moved out of this `if` and https://github.com/percona/pmm-managed/pull/455/files#diff-c2e3f6e9d52d388949d61ae9cfc0bd7bR429-R430 can be removed.

Reply:
If i remove  ```if alertManagerURL == "" {
		return nil
	}``` , parse will raise error.

 And is it correct to rewrite url each time, even if we don't need to do this?

Reply:
I didn't ask to remove `https://github.com/percona/pmm-managed/pull/455/files#diff-c2e3f6e9d52d388949d61ae9cfc0bd7bR431`, I mentioned two lines above this line.

Reply:
Ah, get it, but in this case go-template will fail, because of `"missingkey=error"` is set for templates.

Reply:
changed